async_service/trio.py

```diff
@@ -1,43 +1,86 @@
 import functools
 import sys
-from typing import Any, AsyncIterator, Awaitable, Callable, cast
+from typing import Any, AsyncIterator, Awaitable, Callable, cast, Dict, List, Tuple
 
 from async_generator import asynccontextmanager
 import trio
 import trio_typing
 
-from ._utils import get_task_name
+from ._utils import get_task_name, iter_dag
 from .abc import ManagerAPI, ServiceAPI
 from .base import BaseManager
 from .exceptions import DaemonTaskExit, LifecycleError
 from .typing import EXC_INFO
 
 
 class TrioManager(BaseManager):
-    # A nursery for sub tasks and services.  This nursery is cancelled if the
-    # service is cancelled but allowed to exit normally if the service exits.
-    _task_nursery: trio_typing.Nursery
+    # The nursery running our root task. If any tasks spawn other tasks, we create separate,
+    # nested nurseries for them so that they are terminated in the correct order.  This nursery is
+    # cancelled if the service is cancelled but allowed to exit normally if the service exits.
+    _root_nursery: trio_typing.Nursery
+
+    # A mapping of parent to list of children which tracks the hierarchy of
+    # service tasks.
+    _service_task_dag: Dict[trio.hazmat.Task, List[trio.hazmat.Task]]
+
+    # Individual nurseries for each child task.
+    _task_cancel_scopes: Dict[
+        trio.hazmat.Task,
+        Tuple[trio.CancelScope, trio.Event],
+    ]
 
     def __init__(self, service: ServiceAPI) -> None:
         super().__init__(service)
 
         # events
+        self._cancelled = trio.Event()
         self._started = trio.Event()
         self._stopping = trio.Event()
         self._finished = trio.Event()
 
+        # task dag tracking
+        self._service_task_dag = {}
+        self._task_cancel_scopes = {}
+
+        # child service tracking
+        self._child_managers = []
+
         # locks
         self._run_lock = trio.Lock()
 
     #
     # System Tasks
     #
+    async def _handle_cancelled(self) -> None:
+        """
+        When cancellation is requested this triggers the cancellation of all
+        background tasks.
+        """
+        self.logger.debug("%s: _handle_cancelled waiting for cancellation", self)
+        await self._cancelled.wait()
+        self.logger.debug("%s: _handle_cancelled triggering task DAG cancellation", self)
+
+        for task in iter_dag(self._service_task_dag.copy()):
+            scope, done = self._task_cancel_scopes[task]
+            scope.cancel()
+            await done.wait()
+
+        # Ensure all of the child services that were running will also register
+        # as being cancelled.
+        for child_manager in self._child_managers:
+            child_manager.cancel()
```

Comment:
This is not really necessary, btw. The problem was that the child service in the test was not actually running, so its `.is_cancelled` was silently returning False because the service had no `._root_nursery` attribute. I think we can just modify the test to ensure it runs the child, like in https://github.com/ethereum/async-service/pull/26/commits/d6525ac9a445c7104d71ff84a9786f866889f690#diff-4d028cb98b12f4034fddb928721815a8R189-R193